### DISCUSSION TITLE
Question about zIndex-ing with two containers.

### DISCUSSION THREAD

**Discussion by jt0dd** [pixijs/pixijs#8141] [Q&A] [answered]
My next to-do in my game is to support z-indexing so the character goes behind objects when necessary. I thought it would be as simple as `zIndex`-ing each container with its y index * -1, if that makes sense.

https://www.loom.com/share/cd60a5985b224aa78c362922ab6944cc

Obviously it's not that simple if the character is in a separate parent container from the rest of the world sprites.

I have a character container and a world container separately appended to `app.stage` because my player model stays in the center of the screen. It seemed to make a lot more sense to just have a world container and move the world. The alternative would be to place the character container in the world and every time the world moves, position the player at an inverse offset relative to the movement of the world, but that adds seemingly unnecessary complexity to what seems like could be very simple.

I'm sure plenty of people have brought up this kind of challenge, as it's a fairly standard model for 2D games. Is there some way built into PIXI.js to solve it? Like some way of doing global `zIndex` -ing or having a container not move with its parent?

**Comment by ivanpopelyshev** [comment, marked as answer]
Have you looked at https://github.com/pixijs/layers ?

**Reply by jt0dd** [reply]
I just tried it and had some issues understanding how to use it: https://github.com/pixijs/layers/issues/79

**Reply by domis86** [reply]
@jt0dd i added here some working example on codesandbox: https://github.com/pixijs/layers/issues/79#issuecomment-1024901980 - maybe will help you :)

**Reply by jt0dd** [reply]
Any future PIXIJS-ers who arrive here, this did end up being the best solution for me and as far as I can tell it is highly performant. These guys did a good job on this plugin.

**Comment by bigtimebuddy** [comment]
The zIndex property in Pixi is very simple and local to a Container. Any object with zIndex is sorted with its siblings... and that's it! The zIndex does not work like CSS, as a global render index. 

Using layers (like Ivan suggested), restructuring your depth elements to be flatter, or managing the sorting yourself are probably your options.

**Reply by jt0dd** [reply]
I accept this as fact, but wonder if there's a technical limitation requiring PIXI to work this way. I understand you have a render tree, and containers allow the code user to transform / position / scale one parent and have all the children conveniently follow suit, but I fail to see what fundamental limitation would prevent the rendering engine from solving this problem (avoiding the need for a plugin that might drop FPS or introduce bugs) by introducing a final rendering z-index sorting capability. 

I mean, I'm really trying to think of why it would be hard. The tree defines coordinates of children, which in turn allows the rendering engine to calculate position and transform offsets relative to the parent, and the choice to make the rendering engine force a certain rendering order for the assets based on this structure seems arbitrary. Once you calculate all the positional data, I don't see why you can't allow some handler to re-sort the rendering order for each asset in the scene.

I just need to have a look through the source code I guess.

Would you by chance know which file this is primarily handled by? The tree processing / rendering, I mean.

**Reply by jt0dd** [reply]
For your suggestion of managing the sorting myself: I thought it might be achievable by parsing `app.stage` before each render pass to append all of the sprites in the tree into a single flat container (calculating their properties based on their parent properties, recursively down the tree) and then return it to its original state after the render pass (in order to not have to organize the objects that way in the rest of my code, which would be harder without a hierarchy).

I started on a simple proof of concept, which I thought might work. My expected result of the following code would be that some error gets thrown and rendering fails (since instead of flattening the stage I'm just replacing it with a filler object), but actually there's no effect at all:

```
function flattenStage(stage) {
    return { modified: true }
}

let i = 0
let cache
app.renderer.addListener('prerender', () => {
    i += 1
    cache = app.stage
    app.stage = flattenStage(app.stage)
    if (i % 500 === 0) console.log('prerender app.stage:', app.stage)
})
app.renderer.addListener('postrender', () => {
    app.stage = cache
    if (i % 500 === 0) console.log('postrender app.stage:', app.stage)
})
```

So this outcome (unaffected rendering) would suggest to me that by the time `prerender` is triggered, the stage has already been parsed and thus, I must need to attach to an earlier event. Does such an event exist? I'm scrubbing through the source code but there is a lot of it.

**Reply by jt0dd** [reply]
I more or less ended up spending the day completing a proof of concept of having a pseudo-container that propagates changes through a hierarchy just like containers, but with all the sprites in a flat global container (`app.stage`) for proper z-index control.

The problem is animating a single tile of movement with just 82 sprites in my scene at 42 FPS generates 1207 position change propagations per second. The lag is major attempting a 60FPS tick rate with 1600+ propagations. The library just isn't meant to be controlled like that. Changing the position of each individual element with `.position.set(x,y)` is too inefficient.

So monkey-patching my own pseudo-hierarchy failed. Hopefully I can figure out how to use @ivanpopelyshev 's plugin soon. And hopefully anyone else who ever has this idea sees this thread to save themself the headache of trying.

**Reply by bigtimebuddy** [reply]
You're fighting with PixiJS's display list way too hard. Okay, so looking at your example this is what I see:

* You have some background elements that are not sorted
* You have some foreground elements that need to be depth sorted with a character
* You have the main character who is centered in the camera the whole time

This is what I'd propose for a structure:

* World Container
  * Game Elements Container
    * Player Character Display Object
    * Game Sprite
    * Game Sprite
    * Game Sprite
  * Non-interactive Elements (background) Container 

What we want to do is to sort the Game Elements Container. The easiest way to do that is:

```js
// Sort the game elements every frame by Y-position
// you need to make sure the anchor point is at the foot/base of each element
gameElements.children.sort((a, b) => a.y - b.y);
```

Now for the movement, whenever the main character moves, you would move the world container AND you move the character by the inverse of the same amount:

```js
worldContainer.position.set(100, 100);
playerCharacter.position.set(-100, -100);
```

Doing this should keep the character stationary but move the world. BONUS: I'd also strongly suggest you cull elements that are off screen (visible = false), this will make sure they don't touch the renderer and your performance will be much better!

In Pixi, there are many ways to solve a problem, this is just one way. Using a plugin is another. 

Hope that helps. Good luck.

**Reply by jt0dd** [reply]
One last question about your suggestion please:

- Game Elements Container
-- Player Character Display Object
-- Game Sprite
-- Game Sprite
-- Game Sprite

What if I want to include relative offset with a group of things, let's say an apartment, so I can set position and not need to know global offset. Things like desks in my current model are groups of things which I used containers to group together, maybe a computer, 4 sprites, etc. Then the apartment is another container. It just seemed natural to organize things in such a hierarchy.

So two solutions I imagine:

1. Maybe there's something besides a container I could be using the group things.
2. Maybe I just can't do it do things with local positioning and parent elements like that, just sort of use some wrapper to apply a local offset (but don't update things if the parent moves, just never move the parent)

One situation I can see not working without some kind of hierarchy is a bus. If I want my character to get on a bus or a train and move around, all those elements need to have a parent of some kind so I'm not applying a million positions.

Maybe moving the character's parent so it becomes a child of whatever building / bus / train it's inside of would work, but I kind of suspect that around the edges or at a doorway to the item the rendering would be weird.

**Reply by jt0dd** [reply]
Edit, I did end up going with the plugin and it worked out great for me. Thanks again for your explanations and help regardless.
